index.js

```diff
@@ -1,211 +1,62 @@
-var url = require('url');
-var caw = require('caw');
-var concatStream = require('concat-stream');
-var decompress = require('gulp-decompress');
-var eachAsync = require('each-async');
-var filenamify = require('filenamify');
-var got = require('got');
-var isUrl = require('is-url');
-var objectAssign = require('object-assign');
-var readAllStream = require('read-all-stream');
-var rename = require('gulp-rename');
-var streamCombiner = require('stream-combiner2');
-var PassThrough = require('readable-stream/passthrough');
-var Vinyl = require('vinyl');
-var vinylFs = require('vinyl-fs');
-var Ware = require('ware');
-
-/**
- * Initialize a new `Download`
- *
- * @param {Object} opts
- * @api public
- */
-
-function Download(opts) {
-	if (!(this instanceof Download)) {
-		return new Download(opts);
-	}
-
-	this.opts = objectAssign({encoding: null}, opts);
-	this.ware = new Ware();
-}
-
-module.exports = Download;
-
-/**
- * Get or set URL to download
- *
- * @param {String} url
- * @param {String} dest
- * @api public
- */
-
-Download.prototype.get = function (url, dest) {
-	if (!arguments.length) {
-		return this._get;
-	}
-
-	this._get = this._get || [];
-	this._get.push({
-		url: url,
-		dest: dest
+const fs = require('fs');
+const path = require('path');
+const url = require('url');
+const caw = require('caw');
+const Decompress = require('decompress');
+const filenamify = require('filenamify');
+const getStream = require('get-stream');
+const got = require('got');
+const mkdirp = require('mkdirp');
+const pify = require('pify');
+const fsP = pify(fs);
+
+const createPromise = (uri, output, stream, opts) => new Promise((resolve, reject) => {
+	stream.on('response', res => {
+		const stream = opts.encoding === null ? getStream.buffer(res) : getStream(res, opts);
+
+		stream
+			.then(data => resolve(data))
+			.catch(err => reject(err));
 	});
 
-	return this;
-};
-
-/**
- * Get or set the destination folder
- *
- * @param {String} dir
- * @api public
- */
-
-Download.prototype.dest = function (dir) {
-	if (!arguments.length) {
-		return this._dest;
+	stream.on('error', reject);
+}).then(data => {
+	if (!output) {
+		return data;
 	}
 
-	this._dest = dir;
-	return this;
-};
-
-/**
- * Rename the downloaded file
- *
- * @param {Function|String} name
- * @api public
- */
-
-Download.prototype.rename = function (name) {
-	if (!arguments.length) {
-		return this._rename;
+	if (opts.extract) {
+		const decompress = new Decompress(opts).src(data).dest(path.dirname(output));
+		return pify(decompress.run.bind(decompress))().then(() => data);
 	}
 
-	this._rename = name;
-	return this;
-};
-
-/**
- * Add a plugin to the middleware stack
- *
- * @param {Function} plugin
- * @api public
- */
-
-Download.prototype.use = function (plugin) {
-	this.ware.use(plugin);
-	return this;
-};
-
-/**
- * Run
- *
- * @param {Function} cb
- * @api public
- */
+	return pify(mkdirp)(path.dirname(output))
+		.then(() => fsP.writeFile(output, data))
```

Comment:
Not sure whether this is clear enough, but the last expression is returned, so you could just append `data`:

```js
.then(() => fsP.writeFile(output, data), data)
```